perf/memtx.cc

```diff
@@ -0,0 +1,333 @@
+#include <array>
+#include <cstring>
+#include <random>
+
+#include "box/allocator.h"
+#include "box/box.h"
+#include "box/index_def.h"
+#include "box/iproto_constants.h"
+#include "box/memtx_allocator.h"
+#include "box/memtx_engine.h"
+#include "box/memtx_space.h"
+#include "box/memtx_tx.h"
+#include "box/port.h"
+#include "box/session.h"
+#include "box/space_cache.h"
+#include "box/space_def.h"
+#include "box/tuple.h"
+#include "box/txn.h"
+#include "box/user.h"
+
+#include "core/event.h"
+#include "core/fiber.h"
+#include "core/memory.h"
+
+#include "lua/init.h"
+
+extern "C" {
+#include <lauxlib.h>
+}
+
+#include "rmean.h"
+
+#include <benchmark/benchmark.h>
+
+/**
+ * The memtx singleton encapsulates responsibility for initialization of
+ * everything related to memtx: the engine, the subsystems it depends on, the
+ * spaces and their indexes.
+ */
+class Memtx final {
+public:
+	Memtx(Memtx &other) = delete;
+	Memtx &operator=(Memtx &other) = delete;
+
+	static Memtx &
+	instance()
+	{
+		static Memtx singleton;
+		return singleton;
+	}
+
+	static constexpr std::uint32_t uid = 0;
+	static constexpr std::uint32_t sid = 0;
+	static constexpr std::uint32_t tree_iid = 0;
+	static constexpr std::uint32_t primary_iid = tree_iid;
+
+private:
+	Memtx()
+	{
+		::memory_init();
+		::fiber_init(fiber_c_invoke);
+		::memtx_tx_manager_init();
+		::event_init();
+
+		::space_cache_init();
+		::user_cache_init();
+		::session_init();
+		::tuple_init(nullptr);
+
+		cord()->max_slice.err = TIMEOUT_INFINITY;
+		cord()->slice.err = TIMEOUT_INFINITY;
+
+		tarantool_L = luaL_newstate();
+		if (tarantool_L == nullptr)
+			panic("failed to create new Lua state");
+
+		rmean_box =
+			::rmean_new(iproto_type_strs, IPROTO_TYPE_STAT_MAX);
+		if (rmean_box == nullptr)
+			panic("failed to create new rmean");
+		rmean_error =
+			::rmean_new(rmean_error_strings, RMEAN_ERROR_LAST);
+		if (rmean_error == nullptr)
+			panic("failed to create new rmean");
+
+		memtx = ::memtx_engine_new(".", true, QUOTA_MAX,
```

Comment:
I guess `QUOTA_MAX` is too big for `tuple_arena_max_size`, I get:
```
$ ./perf/memtx.perftest
mapping 4398046511104 bytes for memtx tuple arena...
failed to preallocate 4398046511104 bytes: Cannot allocate memory, check option 'memtx_memory' in box.cfg(..)
```

Reply:
Didn't notice it on my laptop, sorry. Changed it to `1 GiB`, I guess one can tweak it further if necessary.